In [31]:
import sys

In [32]:
list_1 = list(range(50))
print(sys.getsizeof(list_1))
list_2 = list(range(500))
list_3 = list(range(5_000))
list_4 = list(range(5_000_000))

456


In [33]:
list_2 = list(range(500))
print(sys.getsizeof(list_2))

4056


In [34]:
list_3 = list(range(5_000))
print(sys.getsizeof(list_3))

40056


In [35]:
list_4 = list(range(5_000_000))
print(sys.getsizeof(list_4))


40000056


In [36]:
import time

In [37]:
start = time.time()

average_1 = sum(list_1) / len(list_1)
print("Średnia listy 1: " + str(average_1))

end = time.time()
full_time = end - start
print("Czas obliczeń: " + str(full_time))

Średnia listy 1: 24.5
Czas obliczeń: 0.00024008750915527344


In [38]:
start = time.time()

average_2 = sum(list_2) / len(list_2)
print("Średnia listy 2: " + str(average_2))

end = time.time()
full_time = end - start
print("Czas obliczeń: " + str(full_time))

Średnia listy 2: 249.5
Czas obliczeń: 0.00027251243591308594


In [39]:
start = time.time()

average_3 = sum(list_3) / len(list_3)
print("Średnia listy 3: " + str(average_3))

end = time.time()
full_time = end - start
print("Czas obliczeń: " + str(full_time))

Średnia listy 3: 2499.5
Czas obliczeń: 0.0006608963012695312


In [40]:
start = time.time()

average_4 = sum(list_4) / len(list_4)
print("Średnia listy 4: " + str(average_4))

end = time.time()
full_time = end - start
print("Czas obliczeń: " + str(full_time))

Średnia listy 4: 2499999.5
Czas obliczeń: 0.0341033935546875


# Wnioski z działaniach na listach

- Duże listy zajmują również coraz więcej pamięci operacyjnej, ponieważ wszystkie elementy muszą zostać zapisane w pamięci komputera
- Wraz ze wzrostem liczby elementów zużycie pamięci rośnie liniowo podobnie jak czas wykonania programu
- Wraz ze wzrostem liczby elementów listy zwiększa się czas potrzebny na obliczenie średniej
- Dla małych list czas wykonania jest znacznie niższy niż dla list z większą ilością danych
- Obliczanie średniej przy użyciu `sum(lista) / len(lista)` wymaga przejścia przez wszystkie elementy listy
- Wraz ze wzrostem liczby średnia wzrasta liniowo od `24.5` do `2499999.5`, tym większa lista tym więcej dodatkowych `9` zostaje dodanych między `24` a `.5`

# Pomysł na szybsze obliczanie średniej

Dla liczb: `0, 1, 2, ..., n-1` nie trzeba tworzyć całej listy gdyż tworzony jest ciąg arytmetyczny gdzie pierwsza liczba to `0` a ostatnia `n-1`
Z tego powodu średnią można zapisać jako `(pierwszy_element + ostatni_element) / 2`
Na przykład: dla `n = 10` mamy `(0 + 9) / 2` co daje poprawny wynik `4.5`

In [41]:
num1 = 0
last_num = 49

start = time.time()

average = (num1 + last_num) / 2

end = time.time()
full_time = end - start

print(average)
print(full_time)

24.5
0.00047206878662109375


In [42]:
num1 = 0
last_num = 499

start = time.time()

average = (num1 + last_num) / 2

end = time.time()
full_time = end - start

print(average)
print(full_time)

249.5
7.462501525878906e-05


In [43]:
num1 = 0
last_num = 4_999

start = time.time()

average = (num1 + last_num) / 2

end = time.time()
full_time = end - start

print(average)
print(full_time)

2499.5
8.320808410644531e-05


In [44]:
num1 = 0
last_num = 4_999_999

start = time.time()

average = (num1 + last_num) / 2

end = time.time()
full_time = end - start

print(average)
print(full_time)

2499999.5
5.173683166503906e-05


# Wnioski 'nowego' sposobu obliczania średniej
- Używając takiego sposobu nie wymagane jest tworzenie listy
- Z powodu nie tworzenia list zajmowane jest mniej pamięci
- jako że nie jest używana funkcja `sum()` i nie jest potrzebna iteracja po liście to nie zależnie od wielkości liczb obliczenie takiej średniej jest niskie i bardzo podobne między różnymi danymi
- tworzenie listy może mieć sens w sytuacji kiedy wykonywane są również inne działania na tych danych co prowadzi że nie jest potrzebne ponowne definiowanie zmiennych lub lista nie tworzy ciągu arytmetycznego
- własność matematyczna problemu praktycznie nie zużywa dodatkowej pamięci oraz nie zależnie od wielkości ciągu czas obliczeń jest tak na prawdę taka sama